### PR TITLE
karma feature

### PR DESCRIPTION
**Pull request by drvan**
Adds karma feature, allowing the use of `?++` and `?--` to increment and decrement karma values, respectively. The target can be anything, including users and channels. After adding karma, an acknowledgement as well as the new value of karma for that target will be sent.

### CODE REVIEW

File: tests/feature/karma/feature_test.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/jakevoytko/crbot/app"
+	"github.com/jakevoytko/crbot/feature/karma"
+	"github.com/jakevoytko/crbot/testutil"
+)
+
```
Comment (drvan) [inline tests/feature/karma/feature_test.go:11]: I can't think of a compelling reason to add any helper functions to testutil/runner. That being said, my testing feels sparse, and there may be a corner case I'm missing.

File: feature/karma/karmaparser.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	if splitContent[0] != p.GetName() {
+		log.Fatal("KarmaParser.Parse called with non-list command", errors.New("wat"))
+	}
+
+	splitContent = util.CollapseWhitespace(splitContent, 1)
+	targetRegexp := regexp.MustCompile("^[[:alnum:]].*$")
+
+	// Resolve the target of the karma, removing leading @ and anything after #
+	var target string
+	if len(splitContent) > 1 {
```
Comment (drvan) [inline feature/karma/karmaparser.go:58]: I don't love having to check the length here, but if I don't, the parser panics if there is no target given to `?++`.

File: feature/karma/feature.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	return &Feature{
+		featureRegistry: featureRegistry,
+		karmaMap:        karmaMap,
+	}
+}
+
+// Parsers gets the learn feature parsers.
+func (f *Feature) Parsers() []feature.Parser {
+	return []feature.Parser{
+		NewKarmaParser(model.Name_KarmaIncrement, true),
```
Comment (jakevoytko) [inline feature/karma/feature.go:26]: Comment what these do... `true /* increment */`

File: feature/karma/karmaparser.go (hunk 2 above)
Comment (jakevoytko) [reply feature/karma/karmaparser.go:58]: It makes sense that it panics though, given that you're accessing something past the end of the array

File: feature/karma/karmaparser.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// Parse parses the given karma command.
+func (p *KarmaParser) Parse(splitContent []string) (*model.Command, error) {
+	if splitContent[0] != p.GetName() {
+		log.Fatal("KarmaParser.Parse called with non-list command", errors.New("wat"))
+	}
+
+	splitContent = util.CollapseWhitespace(splitContent, 1)
+	targetRegexp := regexp.MustCompile("^[[:alnum:]].*$")
```
Comment (jakevoytko) [inline feature/karma/karmaparser.go:54]: Pull this into a file-wide variable

File: feature/karma/karmaexecutor.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	err = e.karmaMap.Set(command.Karma.Target, strconv.Itoa(newKarma))
+	if err != nil {
+		log.Fatal("Error storing new karma value", err)
+	}
+
+	// Send ack.
+	if command.Karma.Increment {
+		s.ChannelMessageSend(channelID.Format(), fmt.Sprintf(MsgIncrementKarma, command.Karma.Target, command.Karma.Target, newKarma))
```
Comment (jakevoytko) [inline feature/karma/karmaexecutor.go:72]: Check and log the errors that this returns, same below

File: feature/karma/karmaexecutor.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+func (e *KarmaExecutor) PublicOnly() bool {
+	return true
+}
+
+const (
+	MsgIncrementKarma = "%v has been upvoted. %v now has %d karma."
+	MsgDecrementKarma = "%v has been downvoted. %v now has %d karma."
+)
+
+func (e *KarmaExecutor) Execute(s api.DiscordSession, channelID model.Snowflake, command *model.Command) {
```
Comment (jakevoytko) [inline feature/karma/karmaexecutor.go:36]: Add comments to exported methods

File: tests/feature/karma/feature_test.go (hunk 1 above)
Comment (jakevoytko) [reply tests/feature/karma/feature_test.go:11]: yeah, typically I like to assert the state in storage. But I think this use case is so simple that you don't need to.

**Review by jakevoytko — APPROVED** [review]
@drvan++

Just going to add a followup commit that properly dereferences user entities + does the nitpicky stuff, and then I'll push your change out!
